src/main/java/org/opensearch/security/support/HostAndCidrMatcher.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.security.support;
+
+import java.net.InetAddress;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import inet.ipaddr.IPAddressString;
+
+/**
+ * A utility class that performs matching of IP addresses against hostname patterns and CIDR ranges.
+ * This matcher supports both wildcard hostname patterns (e.g., *.example.com) and CIDR notation (e.g., 192.168.1.0/24).
+ */
+public class HostAndCidrMatcher {
+    private static final String IP_HOSTNAME = "ip-hostname";
+    private static final String IP_HOSTNAME_LOOKUP = "ip-hostname-lookup";
+
+    protected final Logger log = LogManager.getLogger(HostAndCidrMatcher.class);
+    private final WildcardMatcher hostMatcher;
+    private final List<IPAddressString> cidrMatchers;
+
+    /**
+     * Constructs a new matcher with the specified host patterns.
+     *
+     * @param hostPatterns A list of patterns that can include both hostname wildcards
+     *                     (e.g., *.example.com) and CIDR ranges (e.g., 192.168.1.0/24).
+     *                     Must not be null.
+     * @throws IllegalArgumentException if hostPatterns is null
+     */
+    public HostAndCidrMatcher(List<String> hostPatterns) {
+        if (hostPatterns == null) {
+            throw new IllegalArgumentException("Host patterns cannot be null");
+        }
+
+        this.hostMatcher = WildcardMatcher.from(hostPatterns);
+        this.cidrMatchers = hostPatterns.stream().map(IPAddressString::new).filter(IPAddressString::isIPAddress).toList();
+    }
+
+    /**
+     * Checks if the provided IP address matches any of the configured CIDR ranges.
+     *
+     * @param address The IP address to check. Can be either IPv4 or IPv6.
+     * @return true if the address matches any configured CIDR range, false otherwise
+     *         or if the address is null
+     */
+    public boolean matchesCidr(InetAddress address) {
+        if (address == null || cidrMatchers == null) {
+            return false;
+        }
+
+        try {
+            IPAddressString addressString = new IPAddressString(address.getHostAddress());
+            return cidrMatchers.stream().anyMatch(cidrAddress -> cidrAddress.contains(addressString));
+        } catch (Exception e) {
+            log.warn("Failed to process IP address {}: {}", address, e.getMessage());
+            return false;
+        }
```

Comment:
I am wondering about the cases which would trigger an exception here. Are these known?

If this is just covers unknown error conditions, I would recommend not to catch exceptions here and returning `false` instead. The reason is that there might be a remote chance that there is an issue hidden in the code that can be exploited by attackers which can then be used to bypass filters.

Reply:
I don't think the exceptions are known, we can thrown an exception here to avoid hidden failures.